components/style/gecko/data.rs

```diff
@@ -113,30 +109,29 @@ impl PerDocumentStyleData {
 }
 
 impl PerDocumentStyleDataImpl {
-    /// Recreate the style data if the stylesheets have changed.
-    pub fn flush_stylesheets(&mut self) {
-        let mut stylist = if self.device_changed || self.stylesheets_changed {
-            Some(Arc::get_mut(&mut self.stylist).unwrap())
-        } else {
-            None
-        };
-
-        if self.device_changed {
-            Arc::get_mut(&mut stylist.as_mut().unwrap().device).unwrap().reset();
-            self.device_changed = false;
-            // Force a stylesheet flush if the device has changed.
-            self.stylesheets_changed = true;
+    /// Reset the device state because it may have changed.
+    ///
+    /// Implies also a stylesheet flush.
+    pub fn reset_device(&mut self) {
+        {
+            let mut stylist = Arc::get_mut(&mut self.stylist).unwrap();
+            Arc::get_mut(&mut stylist.device).unwrap().reset();
         }
+        self.stylesheets_changed = true;
```

Comment:
> We need to reconstruct the style data because different rules may match.

Oh ok. So We should put a RESTYLE_SELF | RESTYLE_DESCENDANTS hint on the root then. 

> We can definitely be more granular in the future, but that mechanism isn't there right now.

My understanding is that the servo architecture already differentiates between "stylesheets changed" and "device changed". What machinery is missing?